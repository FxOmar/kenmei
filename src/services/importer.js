/* eslint-disable import/prefer-default-export */
export const processList = (list) => {
  const lists = {};

  Object.entries(list.series).forEach(([_key, series]) => {
    lists[series.name] = series.manga;
  });

  return lists;
};
/* eslint-enable import/prefer-default-export */
