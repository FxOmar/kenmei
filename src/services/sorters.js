/* eslint-disable camelcase */
export const unread = (entry) => {
  const { last_chapter_read_url, last_chapter_available_url } = entry.links;


  return last_chapter_available_url
    && last_chapter_read_url !== last_chapter_available_url;
};

const titleSort = (entryA, entryB) => {
  const entryATitle = entryA.attributes.title.toLowerCase();
  const entryBTitle = entryB.attributes.title.toLowerCase();

  return entryATitle.localeCompare(entryBTitle);
};
/* eslint-enable camelcase */

const newReleasesSort = (entryA, entryB) => Number(unread(entryB))
  - Number(unread(entryA));

const releasedAtSort = (a, b) => {
  const aReleasedAt = a.attributes.last_released_at;
  const bReleasedAt = b.attributes.last_released_at;

  // Descending order, with null always being the oldest
  return (aReleasedAt === null) - (bReleasedAt === null)
    || -(aReleasedAt > bReleasedAt)
    || +(aReleasedAt < bReleasedAt);
};

export function sortBy(array, sortType, order) {
  let sortedData = [];

  switch (sortType) {
  case 'attributes.last_released_at':
    sortedData = array.sort(releasedAtSort);
    break;
  case 'newReleases':
    sortedData = array.sort(releasedAtSort).sort(newReleasesSort);
    break;
  case 'attributes.title':
    sortedData = array.sort(titleSort);
    break;
  default:
    sortedData = array.sort(newReleasesSort);
  }

  if (order === 'descending') { return sortedData.reverse(); }

  return sortedData;
}
