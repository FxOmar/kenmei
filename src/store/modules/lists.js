import { Message } from 'element-ui';
import * as userTags from '@/services/endpoints/UserTags';
import * as mangaEntries from '@/services/endpoints/v2/manga_entries';

// Can't access getter inside mutations, hence this has to be a plain function
export const getEntryIndex = (state, id) => state.entries.findIndex(
  (e) => e.id === id,
);

const state = {
  tags: [],
  entries: [],
  selectedEntries: [],
  entriesPagy: {},
  statuses: [
    { enum: 1, name: 'Reading' },
    { enum: 2, name: 'On Hold' },
    { enum: 3, name: 'Plan to Read' },
    { enum: 4, name: 'Completed' },
    { enum: 5, name: 'Dropped' },
  ],
  tagsLoading: false,
  entriesLoading: true,
  entriesUpdating: false,
  checkboxToggled: false,
};

const getters = {
  findEntryFromIDs: (state) => (ids) => state.entries.find(
    (entry) => ids.includes(entry.id),
  ),
  selectedEntriesIDs: (state) => state.selectedEntries.map((entry) => entry.id),
  selectionModeActive: (state) => state.selectedEntries.length > 0 && state.checkboxToggled,
};

const actions = {
  async getTags({ commit }) {
    const response = await userTags.index();
    const { status, data } = response;

    return status === 200 ? commit('setTags', data) : Message.error(data.error);
  },
  async getEntries({ commit }, { page, status, tagIDs, searchTerm, sort }) {
    commit('setEntriesLoading', true);

    const response = await mangaEntries.index(
      page, status, tagIDs, searchTerm, sort,
    );

    if (response.status === 200) {
      commit('setEntries', response.data.entries);
      commit('setEntriesPagy', response.data.pagy);
    } else {
      Message.error(response.data.error);
    }

    commit('setEntriesLoading', false);
  },
  async getEntriesPagy({ commit }, { page, status, tagIDs, searchTerm, sort }) {
    const response = await mangaEntries.pagyInfo(
      page, status, tagIDs, searchTerm, sort,
    );

    if (response.status === 200) {
      commit('setEntriesPagy', response.data);
    } else {
      Message.error(response.data.error);
    }
  },
};

const mutations = {
  setTags(state, data) {
    state.tags = data;
  },
  setEntries(state, data) {
    state.entries = data;
    state.selectedEntries = [];
  },
  setSelectedEntries(state, { entries, isCheckbox }) {
    state.checkboxToggled = isCheckbox;
    state.selectedEntries = entries;
  },
  setEntriesPagy(state, data) {
    state.entriesPagy = data;
  },
  addEntry(state, data) {
    state.entries.unshift(data);
  },
  addSelectedEntry(state, { entry, isCheckbox }) {
    state.checkboxToggled = isCheckbox;
    state.selectedEntries.push(entry);
  },
  updateEntry(state, data) {
    state.entries.splice(getEntryIndex(state, data.id), 1, data);
  },
  replaceEntry(state, { currentEntry, newEntry }) {
    state.entries.splice(getEntryIndex(state, currentEntry.id), 1, newEntry);
  },
  setTagsLoading(state, data) {
    state.tagsLoading = data;
  },
  setEntriesLoading(state, data) {
    state.entriesLoading = data;
  },
  setEntriesUpdating(state, data) {
    state.entriesUpdating = data;
  },
};

export default {
  namespaced: true,
  state,
  getters,
  actions,
  mutations,
};
