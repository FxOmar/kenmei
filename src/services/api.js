import { secure } from '@/modules/axios';

export const addMangaEntry = (seriesURL, mangaListID) => secure
  .post('/api/v1/manga_entries/', {
    manga_entry: {
      series_url: seriesURL,
      manga_list_id: mangaListID,
    },
  })
  .then((response) => {
    if (response.data.error) { return {}; }

    return response.data;
  });

export const updateMangaEntry = (id, attributes) => secure
  .put(`/api/v1/manga_entries/${id}`, { manga_entry: attributes })
  .then((response) => {
    if (response.data.error) { return false; }

    return response.data.data;
  })
  .catch(_error => false);

export const bulkUpdateMangaEntry = (ids, attributes) => secure
  .put('/api/v1/manga_entries/bulk_update', { ids, manga_entry: attributes })
  .then((response) => {
    if (response.data.error) { return false; }

    return response.data.data;
  })
  .catch(_error => false);

export const bulkDeleteMangaEntries = seriesIDs => secure
  .delete('/api/v1/manga_entries/bulk_destroy', { data: { ids: seriesIDs } })
  .then(() => true)
  .catch(() => false);
