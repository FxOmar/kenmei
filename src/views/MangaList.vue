<template lang="pug">
  .container.mx-auto.w-full.h-full.flex.flex-col.items-center
    .flex.flex-col.w-full.max-w-4xl.py-16
      .mx-5.mb-5(class="md:mx-0")
        el-button.float-right(
          ref="openAddMangaModalButton"
          type="primary"
          size="medium"
          @click="dialogVisible = true"
          round
        )
          i.el-icon-plus.mr-1
          | Add Manga
        el-button.float-right.mr-3(
          type="success"
          size="medium"
          @click="importDialogVisible = true"
          round
        )
          i.el-icon-upload2.mr-1
          | Import
      .flex-grow.mx-5(class="md:mx-0")
        the-manga-list(:tableData='tableData')
      el-dialog(
        title="Import Manga List"
        :visible.sync="importDialogVisible"
        custom-class="custom-dialog"
        width="400px"
      )
        el-upload(
          ref="upload"
          action=""
          :http-request="processUpload"
          :multiple="false"
          :show-file-list="false"
          accept="application/json"
          drag
          )
          i.el-icon-upload
          .el-upload__text
            | Drop file here or click to upload
          .el-upload__tip(slot="tip")
            | You can download your Trackr.moe list
            |
            el-link.align-baseline.text-xs(
              href="https://trackr.moe/user/options"
              :underline="false"
              target="_blank"
            )
              | here
            progress-bar.mt-2(:percentage='importProgress')
      el-dialog(
        title="Add Manga"
        :visible.sync="dialogVisible"
        custom-class="custom-dialog"
        width="400px"
      )
        label.font-size-b.primary-text MangaDex series URL
        el-input.mt-3(
          v-model="mangaURL"
          placeholder="https://mangadex.org/title/7139/one-punch-man"
        )
        span(slot="footer" class="dialog-footer")
          el-button(@click="dialogVisible = false") Cancel
          el-button(
            ref="addMangaButton"
            type="primary"
            @click="mangaDexSearch"
            :disabled="mangaURL.length === 0"
          )
            | Add
</template>

<script>
  import pLimit from 'p-limit';
  import {
    Message, Loading, Dialog, Button, Input, Upload, Link,
  } from 'element-ui';

  import TheMangaList from '@/components/TheMangaList';
  import ProgressBar from '@/components/ProgressBar';
  import { getManga, extractSeriesID, getMangaBulk } from '@/services/api';
  import Importer from '@/services/importer';

  export default {
    name: 'MangaList',
    components: {
      TheMangaList,
      ProgressBar,
      'el-button': Button,
      'el-dialog': Dialog,
      'el-input': Input,
      'el-upload': Upload,
      'el-link': Link,
    },
    data() {
      return {
        tableData: [],
        mangaURL: '',
        dialogVisible: false,
        importDialogVisible: false,
        importProgress: 0,
      };
    },
    methods: {
      mangaDexSearch() {
        const mangaID = extractSeriesID(this.mangaURL);

        if (this.alreadyExists(mangaID)) {
          Message.info('Manga already added');
          this.dialogVisible = false;
          return;
        }

        const loading = Loading.service({ target: '.el-dialog' });
        getManga(mangaID)
          .then((newManga) => {
            if (Object.keys(newManga).length === 0) {
              Message.info('Manga was not found');
              loading.close();
            } else {
              this.tableData.push(newManga);
              this.closeModal(loading);
            }
          })
          .catch((error) => {
            if (error.response.status === 400) {
              Message.error('URL is incorrect');
              loading.close();
            } else {
              Message.error('Something went wrong');
              this.closeModal(loading);
            }
          });
      },
      closeModal(loading) {
        this.dialogVisible = false;
        loading.close();
        this.mangaURL = '';
      },
      alreadyExists(mangaID) {
        return this.tableData.some(manga => manga.series.url.includes(mangaID));
      },
      sliceIntoBatches(ids) {
        const batchSize = 20;
        const IDChunks = [];

        for (let i = 0; i < ids.length; i += batchSize) {
          IDChunks.push(ids.slice(i, i + batchSize).join(' '));
        }

        return IDChunks;
      },
      filterNewMangaDexEntries(list) {
        return Importer(list).filter(value => !this.alreadyExists(value));
      },
      processMangaDexList(list) {
        let seriesImported = 0;
        const promiseLimit = pLimit(1);
        const seriesIDs = this.filterNewMangaDexEntries(list);

        if (seriesIDs.length === 0) {
          Message.info('Nothing new to import');
          return;
        }

        const IDChunks = this.sliceIntoBatches(seriesIDs);
        const requestList = IDChunks.map(ids => promiseLimit(
          () => getMangaBulk(ids).then(
            (importedList) => {
              seriesImported += importedList.length;
              this.importProgress = Math.floor(
                seriesImported / seriesIDs.length * 100
              );

              return importedList;
            }
          )
        ));

        this.importMangaInBatches(requestList);
      },
      async importMangaInBatches(requestList) {
        const loading = Loading.service({ target: '.el-upload-dragger' });
        await Promise.all(requestList)
          .then((importedList) => {
            const importedManga = importedList.flat();
            importedManga.forEach(newManga => this.tableData.push(newManga));

            Message.info(`Imported ${importedManga.length} series`);
            this.importDialogVisible = false;
          })
          .catch((_error) => {
            Message.error('Something went wrong');
          })
          .finally(() => {
            loading.close();
          });
      },
      processUpload(file) {
        // Reset import progress
        this.importProgress = 0;

        const reader = new FileReader();

        reader.onload = ((theFile) => {
          const json = JSON.parse(theFile.target.result);
          this.processMangaDexList(json);
        });

        reader.readAsText(file.file);
      },
    },
  };
</script>

<style media="screen" lang="scss">
  .el-button.float-right + .el-button.float-right {
    @apply ml-0;
  }

  @media (max-width: 640px) {
    .custom-dialog {
      width: 100% !important;
    }
  }
</style>
