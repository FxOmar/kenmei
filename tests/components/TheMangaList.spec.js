import { Message } from 'element-ui';
import Vuex from 'vuex';
import flushPromises from 'flush-promises';

import MangaList from '@/components/TheMangaList.vue';
import lists from '@/store/modules/lists';
import mangaEntryFactory from '../factories/mangaEntry';

import * as api from '@/services/api';

const localVue = createLocalVue();

localVue.directive('loading', true);
localVue.directive('tippy', true);
localVue.use(Vuex);

describe('TheMangaList.vue', () => {
  let mangaList;
  let store;

  const defaultEntries = mangaEntryFactory.buildList(1);

  beforeEach(() => {
    store = new Vuex.Store({
      modules: {
        lists: {
          namespaced: true,
          state: {
            lists: [],
            entries: [mangaEntryFactory.build({ id: '1' })],
          },
          mutations: lists.mutations,
        },
      },
    });
    mangaList = mount(MangaList, {
      store,
      localVue,
      sync: false,
      propsData: {
        tableData: defaultEntries,
      },
    });
  });

  describe('when updating a manga entry', () => {
    let updateMangaEntryMock;

    beforeEach(() => {
      updateMangaEntryMock = jest.spyOn(api, 'updateMangaEntry');
      mangaList.setData({ sortedData: defaultEntries });
    });

    afterEach(() => {
      jest.restoreAllMocks();
    });

    it('mutates the state and shows success message', async () => {
      const infoMessageMock = jest.spyOn(Message, 'info');
      const mangaEntry = mangaEntryFactory.build({ id: 1 });

      mangaEntryFactory.attributes.title = 'Manga Title';
      mangaEntryFactory.attributes.last_chapter_read = '2';
      mangaEntryFactory.attributes.last_chapter_available = '2';

      updateMangaEntryMock.mockResolvedValue(mangaEntry);

      mangaList.find({ ref: 'updateEntryButton' }).trigger('click');

      await flushPromises();

      expect(infoMessageMock).toHaveBeenCalledWith('Updated last read chapter');
    });

    it('shows error message if update failed', async () => {
      const errorMessageMock = jest.spyOn(Message, 'error');

      // TODO: Change to correct mockRejectedValue, when I am able to fix the
      // issue with using it
      updateMangaEntryMock.mockResolvedValue(false);

      mangaList.find({ ref: 'updateEntryButton' }).trigger('click');

      await flushPromises();

      expect(errorMessageMock).toHaveBeenCalledWith(
        "Couldn't update. Try refreshing the page"
      );
    });
  });

  describe('@events', () => {
    const entry1 = mangaEntryFactory.build({ id: '1' });
    const entry2 = mangaEntryFactory.build({ id: '2' });

    beforeEach(() => {
      mangaList.setData({ sortedData: [entry1, entry2] });
    });

    it('@handleSelectionChange - when selecting rows, emits seriesSelected', async () => {
      mangaList.findAll('.el-checkbox').trigger('click');

      expect(mangaList.emitted('seriesSelected')[1][0]).toEqual([entry1, entry2]);
    });

    it('@editEntry - when editing an entry, emits editEntry', async () => {
      mangaList.find({ ref: 'editEntryButton' }).trigger('click');

      expect(mangaList.emitted('editEntry')).toBeTruthy();
      expect(mangaList.emitted('editEntry')[0]).toEqual([entry2]);
    });
  });

  describe('when no last chapter is available', () => {
    it('Released at column shows Unknown', async () => {
      const entry = mangaEntryFactory.build();

      entry.attributes.title = 'Manga Title';
      entry.attributes.last_released_at = null;

      mangaList.setProps({ tableData: [entry] });

      await flushPromises();

      expect(mangaList.text()).toContain('Unknown');
    });

    it('Latest Chapter column shows no chapters', async () => {
      mangaList.setProps({
        tableData: [
          mangaEntryFactory.build(
            { links: { last_chapter_available_url: null } }
          ),
        ],
      });

      await flushPromises();

      expect(mangaList.text()).toContain('No chapters');
    });
  });

  describe(':props', () => {
    it(':tableData - sanitizes manga title to convert special characters', async () => {
      const entry = mangaEntryFactory.build();

      entry.attributes.title = '&Uuml;bel Blatt';

      mangaList.setProps({ tableData: [entry] });

      await flushPromises();

      expect(mangaList.find('.el-link--inner').text()).toContain('Ãœbel Blatt');
    });

    it(':tableData - shows sites tracked if more than one', async () => {
      const newMangaEntry = mangaEntryFactory.build();

      newMangaEntry.attributes.tracked_entries.push({
        id: 2,
        manga_source_id: 2,
        manga_series_id: 1,
      });

      mangaList.setProps({ tableData: [newMangaEntry] });

      await flushPromises();

      expect(mangaList.text()).toContain('2 sites tracked');
    });
  });
});
