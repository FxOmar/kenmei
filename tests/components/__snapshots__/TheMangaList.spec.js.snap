// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TheMangaList.vue :props :tableData - renders rows 1`] = `
<tbody>
  <tr class="el-table__row">
    <td rowspan="1" colspan="1" class="el-table_1_column_1  el-table-column--selection">
      <div class="cell"><label class="el-checkbox"><span class="el-checkbox__input"><span class="el-checkbox__inner"></span><input type="checkbox" aria-hidden="false" class="el-checkbox__original" value=""></span>
          <!----></label></div>
    </td>
    <td rowspan="1" colspan="1" class="el-table_1_column_2  ">
      <div class="cell"><a href="series.example.url" target="_blank" class="break-normal el-link el-link--default">
          <!----><span class="el-link--inner">Manga Title</span>
          <!----></a></div>
    </td>
    <td rowspan="1" colspan="1" class="el-table_1_column_3  ">
      <div class="cell"><a href="chapter.example.url" target="_blank" class="el-link el-link--default">
          <!----><span class="el-link--inner">10</span>
          <!----></a></div>
    </td>
    <td rowspan="1" colspan="1" class="el-table_1_column_4  ">
      <div class="cell">a year ago</div>
    </td>
  </tr>
  <!---->
</tbody>
`;
